pkg/controller/registry/reconciler/grpc.go

```diff
@@ -628,15 +606,17 @@ func (c *GrpcRegistryReconciler) CheckRegistryServer(logger *logrus.Entry, catal
 	if err != nil {
 		return false, err
 	}
-	current, err := c.currentPodsWithCorrectImageAndSpec(logger, source, serviceAccount, registryPodSecurityConfig)
+	currentPods, err := c.currentPodsWithCorrectImageAndSpec(logger, source, serviceAccount, registryPodSecurityConfig)
 	if err != nil {
 		return false, err
 	}
-	if len(current) < 1 ||
+	if len(currentPods) < 1 ||
 		service == nil || c.currentServiceAccount(source) == nil {
 		return false, nil
 	}
-
+	if deadPodsDetected, e := detectAndDeleteDeadPods(logger, c.OpClient, currentPods, source.GetNamespace()); deadPodsDetected {
```

Comment:
I still think we should return `true, nil` from this function in the case where:
- There is at least one healthy pod
- There are multiple "dead" pods
- We deleted all of the "dead" pods successfully.

That sort of behavior would mean that we would short circuit and avoid calling `EnsureRegistryServer`, which would:
  - end up making some no-op calls to the apiserver:
  	- e.g. https://github.com/anik120/operator-lifecycle-manager/blob/20a171690cf96fcea177e5fa90426e7c234a14d3/pkg/controller/registry/reconciler/grpc.go#L459
  - (maybe, not sure) have some strange issues related to caching, where subsequent calls to `currentPods` would still return the dead pods because the deletion has no propagated back to our informer cache yet (I'm assuming `c.Lister` is backed by a cache. If it is not, then we'd be making more apiserver calls unnecessarily)

Reply:
Essentially, I think both of the following should result in identical behavior:
1. There are 1 or more alive pods, and no dead pods
2. There are 1 ore more alive pods, and we deleted a bunch of dead pods (hence: there are no dead pods, so this is actually a variant of (1))

Reply:
@joelanford thanks for the discussion yesterday. Thought about it a little more and we can do this. I've had to add an additional change though, highlighting it in the next comment...